### PR TITLE
minimum ECC key size check at TLS/SSL level

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**
Set a minimum ECC key size allowed at the TLS/SSL level. Very similar in functionality to previous RSA min key size pull request.

### CODE REVIEW

File: src/internal.c [hunk 1]
```diff
@@ -13982,6 +14006,12 @@ static word32 QSH_KeyExchangeWrite(WOLFSSL* ssl, byte isServer)
                 WOLFSSL_MSG("Using ECC client cert");
                 usingEcc = 1;
                 sigOutSz = MAX_ENCODED_SIG_SZ;
+
+                /* check minimum size of ECC key */
+                if (wc_ecc_size(&eccKey) < ssl->options.minEccKeySz) {
+                    WOLFSSL_MSG("ECC key size too small");
```
Comment (toddouska) [inline src/internal.c]: Compare types should be same sign.

File: src/internal.c [hunk 2]
```diff
@@ -14937,6 +14967,14 @@ int DoSessionTicket(WOLFSSL* ssl,
                                     goto exit_sske;
                                 }
                                 sigSz = wc_ecc_sig_size((ecc_key*)ssl->sigKey);  /* worst case estimate */
+
+                                /* check the minimum ECC key size */
+                                if (wc_ecc_size((ecc_key*)ssl->sigKey) <
+                                        ssl->options.minEccKeySz) {
+                                    WOLFSSL_MSG("ECC key size too small");
```
Comment (toddouska) [inline src/internal.c]: Compare sign.

File: src/internal.c [hunk 3]
```diff
@@ -17219,6 +17257,11 @@ int DoSessionTicket(WOLFSSL* ssl,
                                 ssl->buffers.key->length);
                             if (ret == 0) {
                                 private_key = (ecc_key*)ssl->sigKey;
+                                if (wc_ecc_size(private_key) <
+                                        ssl->options.minEccKeySz) {
+                                    WOLFSSL_MSG("ECC key too small");
```
Comment (toddouska) [inline src/internal.c]: Compare sign.

File: src/ssl.c [hunk 4]
```diff
@@ -3524,6 +3560,23 @@ static int ProcessBuffer(WOLFSSL_CTX* ctx, const unsigned char* buff,
                 wc_ecc_free(&key);
                 return SSL_BAD_FILE;
             }
+
+            /* check for minimum ECC key size and then free */
+            if (ssl) {
+                if (wc_ecc_size(&key) < ssl->options.minEccKeySz) {
```
Comment (toddouska) [inline src/ssl.c]: Compare sign.

File: wolfssl/internal.h [hunk 5]
```diff
@@ -1063,6 +1063,24 @@ enum Misc {
     /* 150 suites for now! */
 #endif
 
+/* set minimum ECC key size allowed */
+#ifndef WOLFSSL_MIN_ECC_BITS
+    #ifdef WOLFSSL_MAX_STRENGTH
+        #define WOLFSSL_MIN_ECC_BITS  256
+    #else
+        #define WOLFSSL_MIN_ECC_BITS 160
```
Comment (toddouska) [inline wolfssl/internal.h]: I think non max strength should be 224 bits default min.

File: wolfssl/test.h [hunk 6]
```diff
@@ -236,6 +236,11 @@
 #else
     #define DEFAULT_MIN_RSAKEY_BITS 1024
 #endif
+#if !defined(NO_FILESYSTEM) && defined(WOLFSSL_MAX_STRENGTH)
+    #define DEFAULT_MIN_ECCKEY_BITS 256
+#else
+    #define DEFAULT_MIN_ECCKEY_BITS 112 /* secp112r1 smallest in wolfSSL */
```
Comment (toddouska) [inline wolfssl/test.h]: I prefer 224 bits for non max strength default min.

File: src/internal.c [hunk 7]
```diff
@@ -5534,6 +5545,16 @@ static int DoCertificate(WOLFSSL* ssl, byte* input, word32* inOutIdx,
                             #endif /* HAVE_ECC */
                         #endif /*HAVE_PK_CALLBACKS */
                     }
+
+                    /* check size of peer ECC key */
+                    if (ret == 0 && ssl->peerEccDsaKeyPresent &&
+                                              !ssl->options.verifyNone &&
+                                              wc_ecc_size(ssl->peerEccDsaKey)
+                                              < ssl->options.minEccKeySz) {
```
Comment (toddouska) [inline src/internal.c]: Isn't this comparing a signed value (wc_ecc_size) to an unsigned value (minEccKeySz)?  If so, that should be avoided.

**Comment by toddouska** [comment]
Please review signs of compares and default min for non max strength.

**Comment by JacobBarthelmeh** [comment]
Thanks for the review Todd, will address the issues highlighted.

**Pull request merged**